File: 0202-happy-number/0202-happy-number.java
class Solution {
    public boolean isHappy(int n) {
        Set<Integer> s = new HashSet<>();

        while (true) {
            n = sumSquareDigits(n);
            if (n == 1)
                return true;
            if (s.contains(n))
                return false;
            s.add(n);
        }
    }

    private int sumSquareDigits(int n) {
        int num = 0;
        while (n != 0) {
            int digit = n % 10;
            num += digit * digit;
            n /= 10;
        }
        return num;
    }
}

File: 0202-happy-number/0202-happy-number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        seen = set()

        while True:
            n = self.sumSquareDigits(n)
            if n == 1:
                return True
            if n in seen:
                return False
            seen.add(n)

    def sumSquareDigits(self, n: int) -> int:
        num = 0
        while n != 0:
            digit = n % 10
            num += digit * digit
            n //= 10
        return num
